x-pack/plugins/saved_objects_tagging/server/services/assignments/assignment_service.ts

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { uniq, difference } from 'lodash';
+import { PublicMethodsOf } from '@kbn/utility-types';
+import {
+  SavedObjectsClientContract,
+  ISavedObjectTypeRegistry,
+  KibanaRequest,
+  SavedObjectsBulkGetObject,
+} from 'src/core/server';
+import { SecurityPluginSetup } from '../../../../security/server';
+import {
+  AssignableObject,
+  UpdateTagAssignmentsOptions,
+  FindAssignableObjectsOptions,
+  getKey,
+  ObjectReference,
+} from '../../../common/assignments';
+import { updateTagReferences } from '../../../common/references';
+import { taggableTypes } from '../../../common/constants';
+import { getUpdatableSavedObjectTypes } from './get_updatable_types';
+import { AssignmentError } from './errors';
+import { toAssignableObject } from './utils';
+
+interface AssignmentServiceOptions {
+  request: KibanaRequest;
+  client: SavedObjectsClientContract;
+  typeRegistry: ISavedObjectTypeRegistry;
+  authorization?: SecurityPluginSetup['authz'];
+}
+
+export type IAssignmentService = PublicMethodsOf<AssignmentService>;
+
+export class AssignmentService {
+  private readonly soClient: SavedObjectsClientContract;
+  private readonly typeRegistry: ISavedObjectTypeRegistry;
+  private readonly authorization?: SecurityPluginSetup['authz'];
+  private readonly request: KibanaRequest;
+
+  constructor({ client, typeRegistry, authorization, request }: AssignmentServiceOptions) {
+    this.soClient = client;
+    this.typeRegistry = typeRegistry;
+    this.authorization = authorization;
+    this.request = request;
+  }
+
+  public async findAssignableObjects({
+    search,
+    types,
+    maxResults = 100,
+  }: FindAssignableObjectsOptions): Promise<AssignableObject[]> {
+    const searchedTypes = types
+      ? types.filter((type) => taggableTypes.includes(type))
+      : taggableTypes;
+    const assignableTypes = await this.getAssignableTypes(searchedTypes);
+
+    // if no provided type was assignable, return an empty list instead of throwing an error
+    if (assignableTypes.length === 0) {
+      return [];
+    }
+
+    const searchFields = uniq(
+      assignableTypes.map(
+        (name) => this.typeRegistry.getType(name)?.management!.defaultSearchField!
```

Comment:
Do we validate elsewhere that these types all have this defined? If not, this could be a easy-to-create bug that wouldn't easily be caught.

Reply:
`uniq` implementation removes all undefined values from the array, so worse case, missing search fields will be ignored.